src/app/clusters/scenes/SceneTableImpl.cpp

```diff
@@ -0,0 +1,680 @@
+    /// there. If the target is not in the table, sets idx to the first empty space
+    /// @param target_scene Storage Id of scene to store
+    /// @param idx Index where target or space is found
+    /// @return CHIP_NO_ERROR if managed to find the target scene, CHIP_ERROR_NOT_FOUND if not found and space left
+    ///         CHIP_ERROR_NO_MEMORY if target was not found and table is full
+    CHIP_ERROR Find(SceneStorageId target_scene, SceneIndex & idx)
+    {
+        SceneIndex firstFreeIdx = kUndefinedSceneIndex; // storage index if scene not found
+        uint8_t index           = 0;
+
+        while (index < kMaxScenePerFabric)
+        {
+            if (scene_map[index] == target_scene)
+            {
+                idx = index;
+                return CHIP_NO_ERROR; // return scene at current index if scene found
+            }
+            if (scene_map[index].mEndpointId == kInvalidEndpointId && firstFreeIdx == kUndefinedSceneIndex)
+            {
+                firstFreeIdx = index;
+            }
+            index++;
+        }
+
+        if (firstFreeIdx < kMaxScenePerFabric)
+        {
+            idx = firstFreeIdx;
+            return CHIP_ERROR_NOT_FOUND;
+        }
+
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    CHIP_ERROR Save(PersistentStorageDelegate * storage) override
+    {
+        ReturnErrorOnFailure(Register(storage));
+        return PersistentData::Save(storage);
+    }
+
+    CHIP_ERROR Delete(PersistentStorageDelegate * storage) override
+    {
+        ReturnErrorOnFailure(Unregister(storage));
+        return PersistentData::Delete(storage);
+    }
+
+    CHIP_ERROR SaveScene(PersistentStorageDelegate * storage, const SceneTableEntry & entry)
+    {
+        CHIP_ERROR err;
+        SceneTableData scene(fabric_index, entry.mStorageId, entry.mStorageData);
+        // Look for empty storage space
+
+        err = this->Find(entry.mStorageId, scene.index);
+
+        if (CHIP_NO_ERROR == err)
+        {
+
+            return scene.Save(storage);
+        }
+
+        if (CHIP_ERROR_NOT_FOUND == err) // If not found, scene.index should be the first free index
+        {
+            scene_count++;
+            scene_map[scene.index] = scene.mStorageId;
+            ReturnErrorOnFailure(this->Save(storage));
+
+            return scene.Save(storage);
+        }
+
+        return CHIP_ERROR_INVALID_LIST_LENGTH;
+    }
+
+    CHIP_ERROR RemoveScene(PersistentStorageDelegate * storage, const SceneStorageId & scene_id)
+    {
+        SceneTableData scene(fabric_index, scene_id);
+        // Look for empty storage space
+
+        VerifyOrReturnError(this->Find(scene_id, scene.index) == CHIP_NO_ERROR, CHIP_ERROR_NOT_FOUND);
+        ReturnErrorOnFailure(scene.Delete(storage));
+
+        if (scene_count > 0)
+        {
+            scene_count--;
+            scene_map[scene.index].Clear();
+            ReturnErrorOnFailure(this->Save(storage));
```

Comment:
Again, if this fails but we already deleted `scene`, do we not end up in an inconsistent state?